AppMaster failed to receover and the Worker will crash

How to reproduce: throw an exception when AppMaster initializing. 
Worker log:
[ERROR] [03/05/2015 15:24:00.987] [OneForOneStrategy] actor name [app1-executor-1] is not unique!
akka.actor.InvalidActorNameException: actor name [app1-executor-1] is not unique!
        at akka.actor.dungeon.ChildrenContainer$NormalChildrenContainer.reserve(ChildrenContainer.scala:130)
        at akka.actor.dungeon.Children$class.reserveChild(Children.scala:76)
        at akka.actor.ActorCell.reserveChild(ActorCell.scala:369)
        at akka.actor.dungeon.Children$class.makeChild(Children.scala:201)
        at akka.actor.dungeon.Children$class.actorOf(Children.scala:37)
        at akka.actor.ActorCell.actorOf(ActorCell.scala:369)
        at org.apache.gearpump.cluster.worker.Worker$$anonfun$appMasterMsgHandler$1.applyOrElse(Worker.scala:106)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:170)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
        at org.apache.gearpump.cluster.worker.Worker.aroundReceive(Worker.scala:51)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
        at akka.actor.ActorCell.invoke(ActorCell.scala:487)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
        at akka.dispatch.Mailbox.run(Mailbox.scala:221)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

--- a/core/src/main/scala/org/apache/gearpump/cluster/master/AppMasterLauncher.scala
+++ b/core/src/main/scala/org/apache/gearpump/cluster/master/AppMasterLauncher.scala
@@ -104,6 +104,7 @@ class AppMasterLauncher(
   def waitForAppMasterToStart(worker : ActorRef, cancel: Cancellable) : Receive = {
     case ActorCreated(appMaster, _) =>
       cancel.cancel()
+      sender ! BindLifeCycle(appMaster)
       LOG.info(s"AppMaster is created, stopping myself...")
       replyToClient(SubmitApplicationResult(Success(appId)))
       context.stop(self)

--- a/core/src/main/scala/org/apache/gearpump/cluster/worker/Worker.scala
+++ b/core/src/main/scala/org/apache/gearpump/cluster/worker/Worker.scala
@@ -38,7 +38,7 @@ import org.apache.gearpump.util._
 import org.slf4j.Logger
 
 import scala.concurrent.duration._
-import scala.concurrent.{ExecutionContext, Future, future}
+import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success, Try}
 
 case class WorkerDescription(workerId: Int, state: String, actorPath: String,
@@ -61,6 +61,7 @@ private[cluster] class Worker(masterProxy : ActorRef) extends Actor with TimeOut
   private var id = -1
   private val createdTime = System.currentTimeMillis()
   private var masterInfo: MasterInfo = null
+  private var executorNameToActor = Map.empty[String, ActorRef]
 
   private var totalSlots: Int = 0
 
@@ -79,7 +80,6 @@ private[cluster] class Worker(masterProxy : ActorRef) extends Actor with TimeOut
   }
 
   private def updateResourceTimeOut(): Unit = {
-
     LOG.error(s"Worker $id update resource time out")
   }
 
@@ -87,10 +87,11 @@ private[cluster] class Worker(masterProxy : ActorRef) extends Actor with TimeOut
     case shutdown @ ShutdownExecutor(appId, executorId, reason : String) =>
       val actorName = ActorUtil.actorNameForExecutor(appId, executorId)
       LOG.info(s"Worker shutting down executor: $actorName due to: $reason")
+      val executorToStop = executorNameToActor.get(actorName)
 
-      if (context.child(actorName).isDefined) {
-        LOG.info(s"Shuttting down child: ${context.child(actorName).get.path.toString}")
-        context.child(actorName).get.forward(shutdown)
+      if (executorToStop.isDefined) {
+        LOG.info(s"Shuttting down child: ${executorToStop.get.path.toString}")
+        executorToStop.get.forward(shutdown)
       } else {
         LOG.info(s"There is no child $actorName, ignore this message")
         sender ! ShutdownExecutorFailed(s"Can not find executor $executorId for app $appId")
@@ -102,7 +103,8 @@ private[cluster] class Worker(masterProxy : ActorRef) extends Actor with TimeOut
       } else {
         val actorName = ActorUtil.actorNameForExecutor(launch.appId, launch.executorId)
 
-        val executor = context.actorOf(Props(classOf[ExecutorWatcher], launch, masterInfo), actorName)
+        val executor = context.actorOf(Props(classOf[ExecutorWatcher], launch, masterInfo))
+        executorNameToActor += actorName ->executor
 
         resource = resource - launch.resource
         allocatedResource = allocatedResource + (executor -> launch.resource)
@@ -144,7 +146,6 @@ private[cluster] class Worker(masterProxy : ActorRef) extends Actor with TimeOut
         }
       }
   }
-
 
   import context.dispatcher
   override def preStart() : Unit = {

--- a/services/src/test/scala/org/apache/gearpump/services/ConfigQueryServiceSpec.scala
+++ b/services/src/test/scala/org/apache/gearpump/services/ConfigQueryServiceSpec.scala
@@ -51,7 +51,7 @@ class ConfigQueryServiceSpec extends FlatSpec with ScalatestRouteTest with Confi
 
   "ConfigQueryService" should "return config for application" in {
     implicit val customTimeout = RouteTestTimeout(15.seconds)
-    (Get("/appConfig/0") ~> appMasterRoute).asInstanceOf[RouteResult] ~> check{
+    (Get("/config/app/0") ~> appMasterRoute).asInstanceOf[RouteResult] ~> check{
       val responseBody = response.entity.asString
       val config = Try(ConfigFactory.parseString(responseBody))
       assert(config.isSuccess)

